credoai/modules/model_modules/security.py

```diff
@@ -0,0 +1,212 @@
+        victim_classifier_acc = sk_metrics.accuracy_score(y_true, y_pred)
+
+        metrics = {
+            'extraction_attack_score': max((thieved_classifier_acc - 0.5) / (victim_classifier_acc - 0.5), 0)
+            }
+        
+        return metrics
+
+    def _get_model(self, input_dim):
+        """Creates a sequential binary classification model 
+
+        Parameters
+        ----------
+        input_dim : int
+            dimension of the feature vector
+        """
+        model = Sequential()
+        model.add(Dense(units=max(int(input_dim/2), 2), input_dim=input_dim, activation='relu'))
+        model.add(Dense(units=max(int(input_dim/4), 2), activation='relu'))
+        model.add(Dense(2, activation='sigmoid'))
+        model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
+        
+        return model
+
+    def _evasion_attack(self, nsamples=100):
+        """Model evasion security attack
+
+        In model evasion, the adversary only has access to the prediction API of a target model
+            which she queries to extract information about the model internals and train a substitute model.
+
+        Returns
+        -------
+        dict
+            Key: evasion_attack_score
+            Value: area under success rate vs median distance threshold curve -- between 0 and 1
+        """
+        hsj = HopSkipJump(classifier=self.victim_model)
+
+        origl_sample = self.x_test[0:nsamples]
+        adver_sample = hsj.generate(origl_sample)
+
+        origl_pred = self.victim_model._model.predict(origl_sample)
+        adver_pred = self.victim_model._model.predict(adver_sample)
+
+        # standardize for robust distance calculation
+        scaler = StandardScaler()
+        scaler.fit(self.x_train)
+        origl_sample_scaled = scaler.transform(origl_sample)
+        adver_sample_scaled = scaler.transform(adver_sample)
+
+        # generate success rate vs median distance threshold curve
+        distances = [np.median(np.linalg.norm(o - a)) for o,a in zip(origl_sample_scaled, adver_sample_scaled)]
+        thresholds = np.linspace(0,1,101)
+        success_rates = []
+        for t in thresholds:
+            idx = np.where(distances <= t)
+            success_rates.append(self._evasion_success_rate(origl_pred[idx], adver_pred[idx]))
+
+        auc = np.trapz(y=success_rates, x=thresholds)
+
+        metrics = {
+            'evasion_attack_score': auc
+            }
+        
+        return metrics
+
+    def _evasion_success_rate(self, l1, l2):
+        """Calculates evasion success rate
+
+        Parameters
+        ----------
+        l1 : list
+            predictions of the original samples
+        l2 : _type_
+            predictions of the adversarial samples
+
+        Returns
+        -------
+        float
+            the proportion of the predictions that have been flipped
+        """
+        if len(l1) > 0:
+            return np.count_nonzero(l1 != l2) / len(l1)
```

Comment:
`l1 == l2` will only work for numpy arrays, but the function signature says lists. Could require numpy arrays or use `np.equal`